File: src/main/java/de/ayont/lpc/channels/BaseChannel.java
package de.ayont.lpc.channels;

import de.ayont.lpc.LPC;
import net.kyori.adventure.text.Component;
import net.kyori.adventure.text.minimessage.MiniMessage;
import net.kyori.adventure.text.serializer.legacy.LegacyComponentSerializer;
import net.luckperms.api.LuckPerms;
import net.luckperms.api.LuckPermsProvider;
import net.luckperms.api.cacheddata.CachedMetaData;
import org.bukkit.Bukkit;
import org.bukkit.configuration.ConfigurationSection;
import org.bukkit.entity.Player;

public abstract class BaseChannel implements Channel {

    protected final LPC plugin;
    protected final String id;
    protected final String format;
    protected final String permission;
    protected final String shortcut;
    protected final String symbol;
    protected final MiniMessage miniMessage;
    protected final LuckPerms luckPerms;

    public BaseChannel(LPC plugin, String id, ConfigurationSection section) {
        this.plugin = plugin;
        this.id = id;
        this.format = section.getString("format", "{prefix}{name} » {message}");
        this.permission = section.getString("permission", "");
        this.shortcut = section.getString("shortcut", "");
        this.symbol = section.getString("symbol", "");
        this.miniMessage = MiniMessage.miniMessage();
        this.luckPerms = LuckPermsProvider.get();
    }

    @Override
    public String getId() {
        return id;
    }

    @Override
    public String getName() {
        return id;
    }

    @Override
    public String getFormat() {
        return format;
    }

    @Override
    public String getPermission() {
        return permission;
    }

    @Override
    public String getShortcut() {
        return shortcut;
    }

    @Override
    public String getSymbol() {
        return symbol;
    }

    @Override
    public boolean canJoin(Player player) {
        return permission.isEmpty() || player.hasPermission(permission);
    }

    @Override
    public boolean canRead(Player player, Player sender) {
        // By default, anyone with join permission can read.
        // Range checks will override this.
        return canJoin(player);
    }

    @Override
    public void sendMessage(Player sender, String message) {
        if (!canJoin(sender)) {
            sender.sendMessage(miniMessage.deserialize(plugin.getConfig().getString("channels.messages.no-permission-speak", "<red>You do not have permission to speak in this channel.")));
            return;
        }

        for (Player recipient : Bukkit.getOnlinePlayers()) {
            if (canRead(recipient, sender)) {
                Component component = plugin.getChatRendererUtil().render(sender, message, recipient, id);
                if (plugin.isPaper()) {
                    recipient.sendMessage(component);
                } else {
                    recipient.sendMessage(LPC.getLegacySerializer().serialize(component));
                }
            }
        }

        // Log to console
        Component consoleComponent = plugin.getChatRendererUtil().render(sender, message, plugin.getAdventure().console(), id);
        if (plugin.isPaper()) {
            plugin.getAdventure().console().sendMessage(consoleComponent);
        } else {
            plugin.getLogger().info(LPC.getLegacySerializer().serialize(consoleComponent));
        }
    }

    @Override
    public void onJoin(Player player) {
        // Optional: Send "You joined X channel"
    }
}

File: src/main/kotlin/de/ayont/lpc/channels/BaseChannel.kt
package de.ayont.lpc.channels

import de.ayont.lpc.LPC
import net.kyori.adventure.text.minimessage.MiniMessage
import net.luckperms.api.LuckPermsProvider
import org.bukkit.Bukkit
import org.bukkit.configuration.ConfigurationSection
import org.bukkit.entity.Player

abstract class BaseChannel(
    protected val plugin: LPC,
    private val id: String,
    section: ConfigurationSection
) : Channel {

    private val format: String = section.getString("format", "{prefix}{name} » {message}")!!
    private val permission: String = section.getString("permission", "")!!
    private val shortcut: String = section.getString("shortcut", "")!!
    private val symbol: String = section.getString("symbol", "")!!
    protected val miniMessage: MiniMessage = MiniMessage.miniMessage()
    private val luckPerms = LuckPermsProvider.get()

    override fun getId(): String = id

    override fun getName(): String = id

    override fun getFormat(): String = format

    override fun getPermission(): String = permission

    override fun getShortcut(): String = shortcut

    override fun getSymbol(): String = symbol

    override fun canJoin(player: Player): Boolean {
        return permission.isEmpty() || player.hasPermission(permission)
    }

    override fun canRead(player: Player, sender: Player): Boolean {
        // By default, anyone with join permission can read.
        // Range checks will override this.
        return canJoin(player)
    }

    override fun sendMessage(sender: Player, message: String) {
        if (!canJoin(sender)) {
            sender.sendMessage(miniMessage.deserialize(plugin.config.getString("channels.messages.no-permission-speak", "<red>You do not have permission to speak in this channel.")!!))
            return
        }

        val renderer = plugin.chatRendererUtil ?: return

        for (recipient in Bukkit.getOnlinePlayers()) {
            if (canRead(recipient, sender)) {
                val component = renderer.render(sender, message, recipient, id)
                if (plugin.isPaper) {
                    recipient.sendMessage(component)
                } else {
                    recipient.sendMessage(LPC.legacySerializer.serialize(component))
                }
            }
        }

        // Log to console
        val console = plugin.adventure?.console() ?: return
        val consoleComponent = renderer.render(sender, message, console, id)
        if (plugin.isPaper) {
            console.sendMessage(consoleComponent)
        } else {
            plugin.logger.info(LPC.legacySerializer.serialize(consoleComponent))
        }
    }

    override fun onJoin(player: Player) {
        // Optional: Send "You joined X channel"
    }
}
